src/Containers/OhmmsPETE/OhmmsArray.h

```diff
@@ -16,100 +16,73 @@
 
 /** A D-dimensional Array class based on PETE
  *
- *  Equivalent to blitz::Array<T,D>, pooma::Array<D,T>.
- *  No operators are provided.
- *  \todo use aligned_vector<T>
  */
-#include <vector>
-#include <iostream>
+#include <array>
+#include <type_traits>
+#include "OhmmsVector.h"
 
-template<class T, unsigned D>
-struct Array
+template<class T, unsigned D, typename ALLOC = std::allocator<T>>
+class Array
 {
+public:
   using Type_t      = T;
-  using Container_t = std::vector<T>;
-  using This_t      = Array<T, D>;
-
-  size_t Length[D];
-  Container_t X;
+  using Container_t = qmcplusplus::Vector<T, ALLOC>;
+  using This_t      = Array<T, D, ALLOC>;
 
   //default constructor
-  Array()
-  {
-    for (int i = 0; i < D; i++)
-      Length[i] = 0;
-  }
+  Array() { Length.fill(0); }
 
   //copy constructor
   Array(const Array& rhs)
   {
-    resize(rhs);
+    resize(rhs.shape());
     std::copy(rhs.begin(), rhs.end(), X.begin());
   }
 
-  // specialized for 1-Dim
-  Array(size_t n) { resize(n); }
-
-  // specialized for 2-Dim
-  Array(size_t n, size_t m) { resize(n, m); }
-
-  // specialized for 3-Dim
-  Array(size_t l, size_t m, size_t n) { resize(l, m, n); }
-
-  // specialized for 4-Dim
-  Array(size_t l, size_t m, size_t n, size_t o) { resize(l, m, n, o); }
-
-  // specialized for something like TinyVector
-  template<typename ST1>
-  Array(ST1* dims)
+  template<typename SIZET = size_t, typename = std::is_integral<SIZET>>
+  Array(const std::array<SIZET, D>& dims)
   {
     resize(dims);
   }
 
-  // do nothing Destructor
-  ~Array() {}
-
   inline unsigned dim() const { return D; }
-  inline size_t* shape() const { return &Length[0]; }
+  inline const std::array<size_t, D>& shape() const { return Length; }
   inline size_t size() const { return X.size(); }
   inline size_t size(int i) const { return Length[i]; }
 
   Container_t& storage() { return X; }
 
-  template<typename TT>
-  void resize(const Array<TT, D>& rhs)
+  template<typename SIZET = size_t, typename = std::is_integral<SIZET>>
+  void resize(const std::array<SIZET, D>& dims)
   {
-    X.resize(rhs.size());
-    for (int i = 0; i < D; i++)
-      Length[i] = rhs.Length[i];
+    for (int i = 0; i < dims.size(); i++)
+      Length[i] = dims[i];
+    X.resize(full_size(Length));
   }
 
+  void resize(size_t n) { resize(std::array<size_t, 1>{n}); }
 
-  template<typename ST1>
-  void resize(ST1* newdims)
-  {
-    int ntot = 1;
-    for (int i = 0; i < D; ++i)
-      ntot *= Length[i] = newdims[i];
-    if (ntot == 0)
-      return;
-    X.resize(ntot);
-  }
+  void resize(size_t m, size_t n) { resize({m, n}); }
 
-  // resize is specialized for D-dimensional
-  void resize(size_t);
-  void resize(size_t, size_t);
-  void resize(size_t, size_t, size_t);
-  void resize(size_t, size_t, size_t, size_t);
+  void resize(size_t l, size_t m, size_t n) { resize({l, m, n}); }
```

Comment:
This is a big improvement, I'd go a little farther make the following change, but it could go in a following PR.

```C++
template<typename... Args>
void resize(Args... sizes)
{
    static_assert(sizeof...(Args) == D);
    resize({static_cast<std::size_t>(std::forward<Args>(sizes))...});
}
```
which then allows the constructor
```C++
/** Provides specialized constructors with signature (size_1, ... , size_D) for  array dimension D
 */
template<typename... Args>
Array(Args... sizes)
{
  resize(sizes...);
}
```